### PR TITLE
Add support for .crypto (UnstoppableDomains) domain lookup #1931

### LINKED ISSUE #1931: Add support for .crypto (UnstoppableDomains) domain lookup
Support New Blockchain name lookup service: https://unstoppabledomains.com/

Lookup is quite straightforward. When presented with a name ending in .crypto:

1. fetch resolver contract, call 'resolverOf(uint256 nameHash)' on contract ```0xD1E5b0FF1287aA9f9A268759062E4Ab08b9Dacbe```

```resolver = function resolverOf(uint256 nameHash) returns Address```

eg ```nameHash of 'brad.crypto' is 0x756e4e998dbffd803c21d23b06cd855cdc7a4b57706c95964a37e24b47c10fc9```
```resolverOf(0x756e4e998dbffd803c21d23b06cd855cdc7a4b57706c95964a37e24b47c10fc9)```

result:

```0xb66DcE2DA6afAAa98F2013446dBCB0f4B0ab2842```

2. lookup name address using the resolver contract found in step 1.

```function get(string key, uint256 nameHash) returns string```

where key is always ```crypto.ETH.address``` for ethereum lookup, eg:

```get('crypto.ETH.address', 0x756e4e998dbffd803c21d23b06cd855cdc7a4b57706c95964a37e24b47c10fc9)```

result:

```0x8aaD44321A86b170879d7A244c1e8d360c99DdA8```

BEWARE! look at the function prototypes, they are different from ENS.

### PR DESCRIPTION
**Pull request by vladyslav-iosdev**
Closes #1931

### CODE REVIEW

**Comment by vladyslav-iosdev** [comment]
@hboon all Trevis checks failed, 
https://travis-ci.com/github/AlphaWallet/alpha-wallet-ios/builds/198257445
there is message 

```
fatal: reference is not a tree: 944b91ba81721eb08705afcfd8d158b0cd42d771
The command "pod install --repo-update" failed and exited with 1 during .
Your build has been stopped.
``` 

Does it mean that i need to call `pod install --repo-update` before send commit?

**Comment by hboon** [comment]
No. Shouldn’t matter. Does the commit hash mentioned exist?

**Review by hboon — CHANGES_REQUESTED** [review]
Are the Podfile changes needed for this PR? If so:

```
pod install
```

> [!] `CryptoSwift` requires CocoaPods version `>= 1.9.1`, which is not satisfied by your current version, `1.8.4`.


Do you want to submit a separate PR to get CocaoPods to 1.9.1 or above?

Also, since Podfile is updated, we need to update `Podfile.lock` too

**Comment by hboon** [comment]
@vladyslav-iosdev remember to buzz me again if this is good for review

**Review by hboon — CHANGES_REQUESTED** [review]
Getting this when `pod install`:

> [!] Error installing TrustKeystore
> [!] Failed to download 'TrustKeystore': [!] /usr/local/bin/git -C /var/folders/x8/s575y2zs4xs6w6ysrvs97dtw0000gq/T/d20201109-81866-1w2vd74 checkout --quiet 0aae827ccfbbcb095bc17753dc75b9bb302392aa
> 
> fatal: reference is not a tree: 0aae827ccfbbcb095bc17753dc75b9bb302392aa

**Comment by vladyslav-iosdev** [comment]
> Getting this when pod install:

its strange, it works for me fine.

**Comment by vladyslav-iosdev** [comment]
@hboon removed cached pods, and reinstall them again, it works for me.

**Comment by hboon** [comment]
@vladyslav-iosdev I tried 2 cases:

A) Install with Cocoapods 1.8.4

> [!] `CryptoSwift` requires CocoaPods version `>= 1.9.1`, which is not satisfied by your current version, `1.8.4`.

B) with Cocoapods 1.9.1

Installs, but looks like we need to check in the `podfile.lock` changes

**Review by hboon — CHANGES_REQUESTED** [review]
@vladyslav-iosdev I tried 2 cases:

A) Install with Cocoapods 1.8.4

> [!] `CryptoSwift` requires CocoaPods version `>= 1.9.1`, which is not satisfied by your current version, `1.8.4`.

B) with Cocoapods 1.9.1

Installs, but looks like we need to check in the `podfile.lock` changes

Can you also pushed PRs (or if pushed, buzz me there) to get the dependent pods' repos merged?

**Comment by vladyslav-iosdev** [comment]
@hboon check it.

**Comment by hboon** [comment]
Looks ok, I noticed you use Cocoapods `1.10.0`. If that's what you intended (and not `1.9.1`), that's fine.

Can you get the PRs for the dependencies? For the `UnstoppableDomainsResolution`, can you submit a PR to their repo and see if we can get it merged? Otherwise we'll have to fork it. Hopefully not. We can still settle the PRs for the other dependencies first.

**Comment by hboon** [comment]
I realised I have forgotten to review the PR itself. Will get to it :)

**Comment by vladyslav-iosdev** [comment]
> Looks ok, I noticed you use Cocoapods `1.10.0`. If that's what you intended (and not `1.9.1`), that's fine.
> 
> Can you get the PRs for the dependencies? For the `UnstoppableDomainsResolution`, can you submit a PR to their repo and see if we can get it merged? Otherwise we'll have to fork it. Hopefully not. We can still settle the PRs for the other dependencies first.

i suppose we will need to make a fork, because i updated UnstoppableDomainsResolution with downgrated dependency version.

**Comment by hboon** [comment]
> > Looks ok, I noticed you use Cocoapods `1.10.0`. If that's what you intended (and not `1.9.1`), that's fine.
> > Can you get the PRs for the dependencies? For the `UnstoppableDomainsResolution`, can you submit a PR to their repo and see if we can get it merged? Otherwise we'll have to fork it. Hopefully not. We can still settle the PRs for the other dependencies first.
> 
> i suppose we will need to make a fork, because i updated UnstoppableDomainsResolution with downgrated dependency version.

I understand. I mean submit a PR to `UnstoppableDomainsResolution` GitHub repo and see if they can merge it in. I presume the change was just to downgrade the dependency and there's no functional change.

**Comment by vladyslav-iosdev** [comment]
i am not sure if it ok for PR, i setted dependency for own forked repo for 
https://github.com/vladyslav-iosdev/EthereumABI.git (it sets CryptoSwift version to 0.10.0)
` s.dependency 'CryptoSwift', '~> 0.10.0'`

**Comment by vladyslav-iosdev** [comment]
i will try to skip these changes (above)

File: AlphaWallet/Core/DomainResolver.swift [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        let node = input.lowercased().nameHash
+        if let value = cachedResult(forNode: node) {
+            return .value(value)
+        }
+
+        do {
+            let resolution = try Resolution(providerUrl: server.rpcURL.absoluteString, network: server.web3NetworkName)
+
+            return Promise { seal in
+                resolution.owner(domain: input) { result in
```
Comment (merenkoff) [inline AlphaWallet/Core/DomainResolver.swift:52]: you need to check not owner. you need check address of ETH wallet. 

```swift 

resolution.addr(domain: "brad.crypto", ticker: "eth") { result in
  switch result {
  case .success(let returnValue):
    // 0x8aaD44321A86b170879d7A244c1e8d360c99DdA8
    let ethAddress = returnValue
  case .failure(let error):
    print("Expected eth Address, but got \(error)")
  }
}

```

**Review by merenkoff — COMMENTED** [review]
Please read our documentation.

**Review by merenkoff — CHANGES_REQUESTED** [review]
Docs https://github.com/unstoppabledomains/resolution-swift#common-examples

**Comment by vladyslav-iosdev** [comment]
@hboon can u review this issue?

**Review by hboon — CHANGES_REQUESTED** [review]
- [ ] I had to change the commit (because the Podfile.lock seems outdated), to finish `pod install`:

```
pod 'UnstoppableDomainsResolution', :git=>'https://github.com/vladyslav-iosdev/resolution-swift.git', :commit=>'9e1283c8deede5ad936e9c355dc92f40d62d54b0'
```

- [ ] Are some of the changes like for `web3swift` and `TrustKeystore` (maybe `CryptoSwift` too?)in `Podfile` because you are mixing in changes required to get the project to build with Xcode12? They don't seem necessary? Can we exclude them

- [ ] Can you create a pull request for the UnstoppableDomainsResolution pod and try to get it merged in? We can still merge this PR with a fork first if the PR hasn't/isn't merge yet. We can update again when that's done

- [x] It doesn't seem to work when I enter "hweeboon.eth"

**Comment by vladyslav-iosdev** [comment]
> * It doesn't seem to work when I enter "hweeboon.eth"

@hboon I updated resolving ens with `GetENSAddressCoordinator` if `UnstoppableDomainsResolution` get failed

**Comment by vladyslav-iosdev** [comment]
> * Are some of the changes like for `web3swift` and `TrustKeystore` (maybe `CryptoSwift` too?)in `Podfile` because you are mixing in changes required to get the project to build with Xcode12? They don't seem necessary? Can we exclude them

yep we can, in case if this PR will be merged after #2143

**Comment by vladyslav-iosdev** [comment]
> * [ ]  Can you create a pull request for the UnstoppableDomainsResolution pod and try to get it merged in? We can still merge this PR with a fork first if the PR hasn't/isn't merge yet. We can update again when that's done

i created PR.

**Comment by hboon** [comment]
> yep we can, in case if this PR will be merged after #2143

Hmm.. ok, let's get #2143 done first. Buzz me if I'm blocking there.

**Review by merenkoff — APPROVED** [review]
--------------------------------------------------------------------------------
 🎉  Congrats

 🚀  UnstoppableDomainsResolution (0.1.6) successfully published
 📅  November 18th, 15:18
 🌎  https://cocoapods.org/pods/UnstoppableDomainsResolution
 👍  Tell your friends!
--------------------------------------------------------------------------------

Please use pod (0.1.6)

**Comment by vladyslav-iosdev** [comment]
@hboon PR is ready for review.

**Comment by hboon** [comment]
> 🎉 Congrats
> 
> ## 🚀 UnstoppableDomainsResolution (0.1.6) successfully published
> 📅 November 18th, 15:18
> 🌎 https://cocoapods.org/pods/UnstoppableDomainsResolution
> 👍 Tell your friends!
> Please use pod (0.1.6)

Thanks for the support! @merenkoff

**Comment by hboon** [comment]
Forced pushed after:

* moved the network name as an `RPCServer` extension along with a comment into `DomainResolver`
* rebased master

**Review by hboon — APPROVED** [review]

**Pull request merged**